File: programmers/17686.py
# python sort 함수는 stable을 보장한다!
# cmp_to_key를 이용한 compare 함수가 동작하지 않는 이유는,,ㅠ => T,F 가 아니라 0, -1, 1 을 반환해야하는건가보다!! 완전 맞는말
# -1 :  현상 유지(앞에것이 앞으로) 
from functools import cmp_to_key

def compare(x, y):
    nx = x[0].lower()
    ny = y[0].lower()
    
    if nx > ny:
        return 1
    elif nx < ny:
        return -1
    else:
        if int(x[1]) > int(y[1]):
            return 1
        elif int(x[1]) < int(y[1]):
            return -1
        else:
            return 0
        
    
def solution(files):
    answer = []

    v = []
    for file in files:
        header = ""
        number = ""
        tail = ""
        
        flag = 0
        
        for c in file:          
            if(c.isdigit()):
                #  아  , , ,,,, digit이기만 하면 1로 바꿔서 문제였다
                if(flag ==0): flag = 1
            elif(flag == 1): 
                flag = 2
                
            if len(number) == 5 : flag = 2
                
            if(flag == 0): header += c
            elif(flag == 1): number += c
            else: tail += c
                
        temp = [header, number, tail]
        v.append(temp)
    
    v = sorted(v, key=cmp_to_key(compare))
    
    for splits in v:
        s = ""
        for split in splits:
            s += split
        answer.append(s)
    return answer

# new version
# python sort 함수는 stable을 보장한다!
# sort key에 lambda 를 활용하자
def solution(files):
    answer = []
    
    for file in files:
        
        head, number, tail = "", "", ""
        flag = False
        
        for i in range(len(file)):
            # number에 추가
            if file[i].isdigit() and len(number) < 5:
                number += file[i]
                flag = True
            # 아직 number가 나오지 않았다면 header
            elif not flag:
                head += file[i]
            # 넘버가 나오고, 숫자도 아니면 나머지는 tail
            else:
                tail += file[i:]
                break
                
        # tuple로 저장
        answer.append((head, number, tail))
            
    # sort , header->number 순으로 와! 이러니까 cmp_to_key를 쓸 필요가 없구나
    answer.sort(key = lambda x : (x[0].upper(), int(x[1])))

    return [''.join(t) for t in answer]

File: programmers/17686.cpp
/*
- 파일명 속 포함된 숫자를 반영한 정렬 기능
- 영문 대소문자, 숫자, 공백, , , - 로 이루어짐
- HEAD(문자), NUMBER, TAIL(나머지) 로 나눠서 각기 비교해야된다.
- files.size() * 3 의 크기면 되겠다. => split 처리 완료
- 커스텀 compare 함수를 넣어서 비교한다.

* 끝까지 같은 경우에는 순서 유지하라고 했는데, c++은 unstable 정렬이라 따로 정렬도 맞춰줘야하는디 , , ,
    => c++ algorithm header에 stable_sort
    
=> 로직은 맞는데, 어디서 오류가 나는걸까~ 아~ 숫자가 5글자면 끝나는거임,,
=> 아~ transform 은 영어외에도 다른문자들이 낑겨있을 수 있음!! 따로 만들어줘야했다.
*/
#include <string>
#include <vector>
#include <algorithm>

#include <iostream>
using namespace std;

string to_lower(string str){
    string ret = "";
    for(int i = 0 ; i < str.size() ; i++){
        if( 'A' <= str[i] && str[i] <= 'Z'){
            ret += str[i] + ('a' - 'A');
        }else ret += str[i];
    }
    return ret;
}
bool compare(vector<string> x, vector<string> y){
    string nx = to_lower(x[0]);
    string ny = to_lower(y[0]);
    
    if(nx != ny){
        return nx < ny;
    }
    return stoi(x[1]) < stoi(y[1]);

}
vector<string> solution(vector<string> files) {
    vector<string> answer;
    
    // files.size() * 3(H, N, T)의 크기의 배열
    vector<vector<string>> v;
    
    // split 해서 집어 넣기
    for(int idx = 0 ; idx < files.size() ; idx++){
        
        string head = "";
        string number = "";
        string tail = "";
        
        int flag = 0;
        
        for(int i = 0 ; i < files[idx].size() ; i++){
            
            // 숫자일 때
            if('0' <= files[idx][i] && files[idx][i] <= '9'){
                if(flag == 0) flag = 1;
                               
            }else if( flag == 1){
                // 숫자가 아닌데, 이미 숫자를 보았을 때 
                flag = 2;
            }
            
            if(number.size() == 5) flag = 2;
            
            if(flag == 0) head += files[idx][i];
            else if(flag == 1) number += files[idx][i];
            else tail += files[idx][i];                
        }
        
        //cout << head << " " << number << " " << tail << endl;
        vector<string> temp;
        temp.push_back(head);
        temp.push_back(number);
        temp.push_back(tail);
        
        v.push_back(temp);
    }
    
    stable_sort(v.begin(), v.end(), compare);
    
    for(int i = 0 ; i < v.size() ; i++){
        vector<string> split = v[i];
        string s = "";
        for(int j = 0 ; j < split.size() ; j++){
            s += split[j];
        }
        answer.push_back(s);
    }
    return answer;
}